lightning/src/blinded_path/payment.rs

```diff
@@ -150,6 +164,83 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 	utils::construct_blinded_hops(secp_ctx, pks, tlvs, session_priv)
 }
 
+/// `None` if underflow or overflow occurs.
+fn amt_to_forward_msat(inbound_amt_msat: u64, payment_relay: &PaymentRelay) -> Option<u64> {
+	let amt_to_fwd_numerator = inbound_amt_msat.checked_sub(payment_relay.fee_base_msat as u64)
+		.and_then(|a| a.checked_mul(1_000_000))
+		.and_then(|a| a.checked_add(1_000_000 - 1))
+		.and_then(|a| a.checked_add(payment_relay.fee_proportional_millionths as u64));
+	let amt_to_fwd_denominator = payment_relay.fee_proportional_millionths.checked_add(1_000_000);
+	let mut amt_to_forward = match (amt_to_fwd_numerator, amt_to_fwd_denominator) {
+		(Some(num), Some(denom)) => num / denom as u64,
+		_ => return None,
+	};
+	let fee_opt = amt_to_forward.checked_mul(payment_relay.fee_proportional_millionths as u64)
+		.and_then(|prop| (prop / 1000000).checked_add(payment_relay.fee_base_msat as u64));
+	let fee = if let Some(f) = fee_opt { f } else { return None };
+	if inbound_amt_msat - fee < amt_to_forward {
+		// Rounding up the forwarded amount resulted in underpaying this node, so take an extra 1 msat
+		// in fee to compensate.
+		amt_to_forward -= 1;
```

Comment:
(Note that this method is reused in #2540 by intermediate nodes to calculate their `amt_to_forward`).

So this part is really weird, but I had to add it because otherwise the `amt_to_forward` formula will sometimes spit out a value that undercharges the relay fee. But we can't *always* round down the `amt_to_forward` or else the recipient will sometimes be underpaid.

Reply:
Aren't other implementations at risk of calculating the forwarding fee (ie the `channel.rs` check) the same way we do, and spuriously failing to forward?

Reply:
I'm just saying if we have to do this weird decrementing thing it seems like other people may also have to and we could share the constants that trigger the spurious failure. Or does this mean we don't calculate the fee in `channel.rs` `htlc_satisfies_config` correctly?